### PR TITLE
Implemented 3D Textures

### PR DESCRIPTION
**Pull request by FernandoS27**
This PR is the culmination of 1 week of hardwork. I was running crazy trying to figure how some games assembled 3D Textures.  I tried everything, asserted copy engine, 3D Framebuffers, etc. And at the end, it was the game rendering a bunch of 2D textures overlapping each other in a region of memory which it later reinterpreted as a 3D Texture.

smo3d3
smo3d2
smo3d1

### CODE REVIEW

File: src/video_core/rasterizer_cache.h [hunk 1]
```diff
@@ -60,6 +60,22 @@ class RasterizerCache : NonCopyable {
         return nullptr;
     }
 
+    std::vector<T> GetInRange(Tegra::GPUVAddr addr, size_t size) {
+        std::vector<T> objects;
```
Comment (lioncash) [inline src/video_core/rasterizer_cache.h]: This can be moved below the if statement, closer to its main usage site, and the if statement can just be changed to:

```cpp
if (size == 0) {
    return {};
}
```

File: src/video_core/rasterizer_cache.h [hunk 2]
```diff
@@ -60,6 +60,22 @@ class RasterizerCache : NonCopyable {
         return nullptr;
     }
 
+    std::vector<T> GetInRange(Tegra::GPUVAddr addr, size_t size) {
```
Comment (lioncash) [inline src/video_core/rasterizer_cache.h]: This can likely be a const member function.

File: src/video_core/renderer_opengl/gl_rasterizer_cache.cpp [hunk 3]
```diff
@@ -1105,6 +1106,83 @@ void RasterizerCacheOpenGL::FermiCopySurface(
...
+                                     dst_params.block_height, dst_params.block_depth);
+
+    glBindBuffer(GL_PIXEL_PACK_BUFFER, 0);
+    glBindBuffer(GL_PIXEL_UNPACK_BUFFER, 0);
+
+    const GLsizei width{static_cast<GLsizei>(dst_params.GetRect().GetWidth())};
+    const GLsizei height{static_cast<GLsizei>(dst_params.GetRect().GetHeight())};
+
+    if (dest_format.compressed) {
+        LOG_CRITICAL(HW_GPU, "Compressed copy is unimplemented!");
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_rasterizer_cache.cpp]: Shouldn't this just use `UNIMPLEMENTED_MSG` instead of the two macros separately?

File: src/video_core/renderer_opengl/gl_rasterizer_cache.cpp [hunk 4]
```diff
@@ -1105,6 +1106,83 @@ void RasterizerCacheOpenGL::FermiCopySurface(
...
+                                dest_format.format, dest_format.type, gl_buffer.data());
+            break;
+        case SurfaceParams::SurfaceTarget::Texture3D:
+        case SurfaceParams::SurfaceTarget::Texture2DArray:
+            glTextureSubImage3D(dst_surface->Texture().handle, 0, 0, 0, 0, width, height,
+                                static_cast<GLsizei>(dst_params.depth), dest_format.format,
+                                dest_format.type, gl_buffer.data());
+            break;
+        default:
+            LOG_CRITICAL(Render_OpenGL, "Unimplemented surface target={}",
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_rasterizer_cache.cpp]: Ditto

File: src/video_core/renderer_opengl/gl_rasterizer_cache.cpp [hunk 5]
```diff
@@ -1105,6 +1106,83 @@ void RasterizerCacheOpenGL::FermiCopySurface(
...
+    const u32 src_bytes_per_pixel = src_params.GetFormatBpp() / 8;
+    const u32 dst_bytes_per_pixel = dst_params.GetFormatBpp() / 8;
+
+    const u32 dst_size = Tegra::Texture::CalculateSize(
+        dst_params.is_tiled, dst_bytes_per_pixel, dst_params.width, dst_params.height,
+        dst_params.depth, dst_params.block_height, dst_params.block_depth);
+
+    std::vector<Surface> surfaces = GetInRange(src_params.addr, dst_size);
+
+    std::sort(surfaces.begin(), surfaces.end(), [](const Surface& a, const Surface& b) -> bool {
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_rasterizer_cache.cpp]: `-> bool` isn't necessary here. It's deducible as a boolean based off the expression alone.

File: src/video_core/renderer_opengl/gl_rasterizer_cache.cpp [hunk 6]
```diff
@@ -1105,6 +1106,83 @@ void RasterizerCacheOpenGL::FermiCopySurface(
...
+
+    Tegra::Texture::CopySwizzledData(dst_params.width, dst_params.height, dst_params.depth,
+                                     dst_bytes_per_pixel, dst_bytes_per_pixel,
+                                     tmp_buffer.data(), gl_buffer.data(), true,
+                                     dst_params.block_height, dst_params.block_depth);
+
+    glBindBuffer(GL_PIXEL_PACK_BUFFER, 0);
+    glBindBuffer(GL_PIXEL_UNPACK_BUFFER, 0);
+
+    const GLsizei width{static_cast<GLsizei>(dst_params.GetRect().GetWidth())};
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_rasterizer_cache.cpp]: These can be `const auto` since the cast already provides the type.

File: src/video_core/renderer_opengl/gl_rasterizer_cache.cpp [hunk 7]
```diff
@@ -1106,6 +1106,83 @@ void RasterizerCacheOpenGL::FermiCopySurface(
...
+    auto dest_format = GetFormatTuple(dst_params.pixel_format, dst_params.component_type);
+
+    const u32 src_bytes_per_pixel = src_params.GetFormatBpp() / 8;
+    const u32 dst_bytes_per_pixel = dst_params.GetFormatBpp() / 8;
+
+    const u32 dst_size = Tegra::Texture::CalculateSize(
+        dst_params.is_tiled, dst_bytes_per_pixel, dst_params.width, dst_params.height,
+        dst_params.depth, dst_params.block_height, dst_params.block_depth);
+
+    std::vector<Surface> surfaces = GetInRange(src_params.addr, dst_size);
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_rasterizer_cache.cpp]: This could use a comment, why does this happen and what are the steps to mitigate the issue explained in plain english

File: src/video_core/renderer_opengl/gl_rasterizer_cache.cpp [hunk 8]
```diff
@@ -1106,6 +1106,8 @@ void RasterizerCacheOpenGL::FermiCopySurface(
     FastCopySurface(GetSurface(src_params, true), GetSurface(dst_params, false));
 }
 
+#pragma optimize("", off)
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_rasterizer_cache.cpp]: umm

File: src/video_core/renderer_opengl/gl_rasterizer_cache.cpp [hunk 9]
```diff
@@ -1124,40 +1126,41 @@ void RasterizerCacheOpenGL::ReinterpretSurface(const Surface& src_surface,
...
 
-    std::sort(surfaces.begin(), surfaces.end(), [](const Surface& a, const Surface& b) -> bool {
-        return a->GetAddr() > b->GetAddr();
-    });
+    std::sort(surfaces.begin(), surfaces.end(),
+              [](const Surface& a, const Surface& b) { return a->GetAddr() > b->GetAddr(); });
 
-    std::vector<u8> tmp_buffer(dst_size);
+    // Temporal counter measure
+    std::vector<u8> tmp_buffer(2 * dst_size);
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_rasterizer_cache.cpp]: Why?

Comment (FernandoS27) [reply src/video_core/renderer_opengl/gl_rasterizer_cache.cpp]: That's workaround a cache bug. Sometimes I'll find surfaces that are supposed to be dead and they cover areas that go out-bounds.

File: src/video_core/renderer_opengl/gl_rasterizer_cache.cpp (hunk 8 above)
Comment (FernandoS27) [reply src/video_core/renderer_opengl/gl_rasterizer_cache.cpp]: oopsies

File: src/video_core/renderer_opengl/gl_rasterizer_cache.cpp (hunk 7 above)
Comment (FernandoS27) [reply src/video_core/renderer_opengl/gl_rasterizer_cache.cpp]: which issue are you talking about?
